File: src/features/typing/data/java/binary_tree_maximum_path_sum.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private int best;

    public int maxPathSum(TreeNode root) {
        best = root.val;
        gain(root);
        return best;
    }

    private int gain(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int l = Math.max(gain(node.left), 0);
        int r = Math.max(gain(node.right), 0);
        if (node.val + l + r > best) {
            best = node.val + l + r;
        }
        return node.val + Math.max(l, r);
    }
}

File: src/features/typing/data/dart/binary_tree_maximum_path_sum.dart
import 'dart:math';

class Solution {
  int best = 0;

  int maxPathSum(TreeNode? root) {
    best = root!.val;
    gain(root);
    return best;
  }

  int gain(TreeNode? node) {
    if (node == null) {
      return 0;
    }
    int l = max(gain(node.left), 0);
    int r = max(gain(node.right), 0);
    if (node.val + l + r > best) {
      best = node.val + l + r;
    }
    return node.val + max(l, r);
  }
}
